h2/utilities.py

```diff
@@ -214,6 +217,22 @@ def _reject_uppercase_header_fields(headers, hdr_validation_flags):
         yield header
 
 
+def _reject_surrounding_whitespace(headers, hdr_validation_flags):
+    """
+    Raises a ProtocolError if any header name or value is surrounded by
+    whitespace characters.
+    """
+    for header in headers:
+        if header[0].strip() != header[0]:
+            raise ProtocolError(
+                "Received header name surrounded by whitespace %s" % header[0])
+        if header[1].strip() != header[1]:
+            raise ProtocolError(
+                "Received header value surrounded by whitespace %s" % header[1]
+            )
+        yield header
```

Comment:
Nitpicking in the extreme, but do we want `%r` instead of `%s` as the format specifiers?

It’s much easier to see the trailing whitespace (and how much of it there is) if the entire name/value is surrounded by quote marks.